gpytorch/lazy/lazy_tensor.py

```diff
@@ -731,6 +731,106 @@ def add_jitter(self, jitter_val=1e-3):
         diag = torch.tensor(jitter_val, dtype=self.dtype, device=self.device)
         return self.add_diag(diag)
 
+    def cat_rows(self, fant_train, fant_fant):
+        """
+        Concatenates new rows and columns to the matrix that this LazyTensor represents, e.g.
+        C = ((A; B); B', D).
+
+        TODO: add reference.
+        """
+        from .cat_lazy_tensor import CatLazyTensor
+        from .root_lazy_tensor import RootLazyTensor
+
+        batch_shape = fant_train.shape[:-2]
+        L = self.root_decomposition().root.evaluate()
+
+        L_inverse = self.root_inv_decomposition().root.evaluate()
```

Comment:
Why do things with dense tensors?

Reply:
These need to be dense roots so that the matrix multiplications in L751 below is okay. In general, fant_train will probably be dense.